apps/arweave/src/ar_tx_queue.erl

```diff
@@ -0,0 +1,284 @@
+			not_found -> ?NUM_EMITTER_PROCESSES;
+			X -> X
+		end,
+	{ok, #state {
+		tx_queue = priority_queue:new(),
+		emitters_running = 0,
+		max_emitters = MaxEmitters,
+		max_size = ?TX_QUEUE_SIZE_LIMIT,
+		size = 0,
+		paused = false,
+		emit_map = #{}
+	}}.
+
+handle_call({set_pause, PauseOrNot}, _From, State) ->
+	{reply, ok, State#state { paused = PauseOrNot }};
+
+handle_call({set_max_emitters, N}, _From, State) ->
+	{reply, ok, State#state { max_emitters = N }};
+
+handle_call({set_max_size, Bytes}, _From, State) ->
+	{reply, ok, State#state { max_size = Bytes }};
+
+handle_call(show_queue, _From, State = #state { tx_queue = Q }) ->
+	Reply = show_queue(Q),
+	{reply, Reply, State};
+
+handle_call(start_emitters, _From, State) ->
+	#state { max_emitters = MaxEmitters, emitters_running = EmittersRunning } = State,
+	lists:foreach(
+		fun(N) ->
+			Wait = N * ?EMITTER_START_WAIT,
+			erlang:send_after(Wait, ?MODULE, emitter_go)
+		end,
+		lists:seq(1, max(MaxEmitters - EmittersRunning, 0))
+	),
+	{reply, ok, State#state { emitters_running = MaxEmitters, paused = false }};
+
+handle_call(_Request, _From, State) ->
+	{noreply, State}.
+
+handle_cast({add_tx, TX}, State) ->
+	#state { tx_queue = Q, max_size = MaxSize, size = Size} = State,
+	TXSize = ?TX_SIZE_BASE + byte_size(TX#tx.data),
+	U = utility(TX, TXSize),
+	{NewQ, NewSize, DroppedTXs} =
+		maybe_drop(
+			priority_queue:in({TX, TXSize}, U, Q),
+			Size + TXSize,
+			MaxSize
+		),
+	case DroppedTXs of
+		[] -> noop;
+		_ -> ar_bridge:drop_waiting_txs(whereis(http_bridge_node), DroppedTXs)
+	end,
+	{noreply, State#state { tx_queue = NewQ, size = NewSize }};
+
+handle_cast(_Request, State) ->
+	{noreply, State}.
+
+handle_info(emitter_go, State = #state { paused = true }) ->
+	erlang:send_after(?EMITTER_START_WAIT, ?MODULE, emitter_go),
+	{noreply, State};
+handle_info(emitter_go, State = #state { emitters_running = EmittersRunning, max_emitters = MaxEmitters }) when EmittersRunning > MaxEmitters ->
+	{noreply, State#state { emitters_running = EmittersRunning - 1}};
+handle_info(emitter_go, State = #state { tx_queue = Q, size = Size, emit_map = EmitMap }) ->
+	NewState =
+		case priority_queue:out(Q) of
+			{empty, _} ->
+				erlang:send_after(?EMITTER_START_WAIT, ?MODULE, emitter_go),
+				State;
+			{{value, {TX, TXSize}}, NewQ} ->
+				Bridge = whereis(http_bridge_node),
+				Peers = lists:sublist(ar_bridge:get_remote_peers(Bridge), ar_meta_db:get(max_propagation_peers)),
+				case Peers of
+					[] ->
+						erlang:send(?MODULE, {emitter_finished, TX}),
+						State#state { tx_queue = NewQ, size = Size - TXSize };
+					_ ->
+						%% Send transactions to the "max_propagation_peers" best peers,
+						%% ?TX_PROPAGATION_PARALLELIZATION peers at a time. The resulting
+						%% maximum for outbound connections is ?TX_PROPAGATION_PARALLELIZATION * num_emitters.
+						lists:foreach(
+							fun(_) ->
+								erlang:send(?MODULE, {emit_tx_to_peer, TX})
```

Comment:
What about `self() ! {emit_tx_to_peer, TX}` here?
I find the `!` operator more concise than `erland:send/2` and I prefer using `self()` inside a gen_server, since it makes the internals independent of whether or not the gen_server has a registered name.